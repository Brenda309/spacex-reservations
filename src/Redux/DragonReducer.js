// API Url
const baseUrl = 'https://api.spacexdata.com/v3/dragons';

// Actions
const GETDRAGONS = 'spacex-reservation/dragons/GET_DRANGONS';
const RESERVE = 'spacex-reservation/dragons/RESERVE';

// Initial state
const initialState = [];

// Action creators
export const getDragonsAPI = () => async (dispatch) => {
  const response = await fetch(baseUrl, {
    method: 'GET',
    headers: {
      'Content-type': 'application/json; charset=UTF-8',
    },
  });
  const dataAPI = await response.json();
  const dragonArr = dataAPI.map((dragon) => {
    const newDragon = {
      dragon_id: dragon.id,
      dragon_name: dragon.name,
      dragon_type: dragon.type,
      dragon_description: dragon.description,
      dragon_images: dragon.flickr_images[2],
      reserved: false,
    };
    return newDragon;
  });
  dispatch({
    type: GETDRAGONS,
    payload: dragonArr,
  });
};

export const reserve = (dragonId) => ({
  type: RESERVE,
  payload: dragonId,
});

// Reducer
export default function dragonsReducer(state = initialState, action) {
  switch (action.type) {
    case GETDRAGONS:
      return action.payload;
    case RESERVE:
      return [
        ...state.map((dragon) => {
          if (dragon.dragon_id === action.payload) {
            return { ...dragon, reserved: true };
          }
          return dragon;
        }),
      ];
    default:
      return state;
  }
}
