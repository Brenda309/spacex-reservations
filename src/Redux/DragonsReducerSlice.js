import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

// API Url
const baseUrl = 'https://api.spacexdata.com/v3/dragons';

// Actions

const GETDRAGONS = 'spacex-reservation/dragons/GET_DRANGONS';

// Async action creators

export const getDragonsAPI = createAsyncThunk(GETDRAGONS, async () => {
  const response = await fetch(baseUrl, {
    method: 'GET',
    headers: {
      'Content-type': 'application/json; charset=UTF-8',
    },
  });

  return response.json();
});

// Slice reducer
export const initialState = [];

const dragonSlice = createSlice({
  name: 'dragons',
  initialState,
  reducers: {
    dragonReserve: {
      reducer: (state, action) => state.map((dragon) => (
        dragon.dragon_id === action.payload ? { ...dragon, reserved: true } : dragon)),
      prepare: (dragonId) => ({ payload: dragonId }),
    },
  },
  extraReducers: {
    [getDragonsAPI.fulfilled]: (state, action) => {
      const dragons = [];
      action.payload.forEach((dragon) => {
        const data = {
          dragon_id: dragon.id,
          dragon_name: dragon.name,
          dragon_type: dragon.type,
          dragon_description: dragon.description,
          dragon_images: dragon.flickr_images[2],
          reserved: false,
        };
        dragons.push(data);
      });
      return dragons;
    },
  },
});

export const { dragonReserve } = dragonSlice.actions;

export default dragonSlice.reducer;
