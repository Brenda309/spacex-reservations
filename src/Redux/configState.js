import { configureStore } from '@reduxjs/toolkit';
import dragonSlice from './DragonsReducerSlice';
import { rocketReducer } from './RocketAPIReducer';
import reducerMissions from './mission';

const rootReducer = reducerMissions;

const Store = configureStore({
  reducer: {
    missions: rootReducer,
    rockets: rocketReducer,
    dragons: dragonSlice,
  },
});

export default Store;
