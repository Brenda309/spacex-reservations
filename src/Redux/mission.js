const FETCH_API = 'spacex-reservations/missions/FETCH_API';
const initialState = [];
const baseMissionsUrl = 'https://api.spacexdata.com/v3/missions';

export const fetchMissions = () => async (dispatch) => {
  const connect = await fetch(baseMissionsUrl);
  const res = await connect.json().then((dataObject) => dataObject);
  const dataArray = res.map((item) => {
    const data = {
      mission_id: item.mission_id,
      mission_name: item.mission_name,
      description: item.description,
      reserved: false,
    };
    return data;
  });
  dispatch({
    type: FETCH_API,
    payload: dataArray,
  });
};

export default function reducerMissions(state = initialState, action) {
  switch (action.type) {
    case FETCH_API:
      return action.payload;
    default:
      return state;
  }
}
