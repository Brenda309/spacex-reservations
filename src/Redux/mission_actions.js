export const GET_MISSION = 'Get_MISSION';
const MISSION_API_URL = 'https://api.spacexdata.com/v3/missions';

export const getMissions = () => {
  try {
    return async (dispatch) => {
      const result = await fetch(MISSION_API_URL, {
        method: 'GET',
        headers: {
          'Constent-Type': 'application/json',
        },
      });
      const json = await result.json();
      if (json) {
        dispatch({
          type: GET_MISSION,
          payload: json,
        });
      } else {
        console.log('unable to fetch!');
      }
    };
  } catch (error) {
    console.log(error);
  }
};
