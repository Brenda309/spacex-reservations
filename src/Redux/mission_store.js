import { configureStore } from '@reduxjs/toolkit';

import postReducer from './missions_reducers';

export default configureStore({
  reducer: {
    post: postReducer,
  },
});
