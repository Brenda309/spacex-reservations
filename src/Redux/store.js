import { configureStore } from '@reduxjs/toolkit';
import thunk from 'redux-thunk';
import reducerMissions from './mission';

const rootReducer = reducerMissions;

const store = configureStore({
  reducer: {
    missions: rootReducer,
  },
  applyMiddleware: [thunk],
});

export default store;
