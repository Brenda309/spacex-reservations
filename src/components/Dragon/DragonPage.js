import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { getDragonsAPI } from '../../Redux/DragonReducer';

import Navigation from '../Navigation/Navigation';
import DragonElement from '../dragonElement/DragonElement';

import styles from './DragonPage.module.scss';

function DragonPage() {
  const dragons = useSelector((state) => state.dragons);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!dragons.length) {
      dispatch(getDragonsAPI());
    }
  }, []);

  return (
    <div>
      <Navigation />
      <main className={styles.container}>
        <section>
          {dragons.map((dragon) => (
            <DragonElement
              key={dragon.dragon_id}
              id={dragon.dragon_id}
              name={dragon.dragon_name}
              type={dragon.dragon_type}
              description={dragon.dragon_description}
              images={dragon.dragon_images}
            />
          ))}
        </section>
      </main>
    </div>
  );
}

export default DragonPage;
