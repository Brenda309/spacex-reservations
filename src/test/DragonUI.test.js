import * as actions from '../Redux/DragonReducer';
import {
  render, screen, waitFor,
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../Redux/configState';
import DragonPage from '../components/Dragon/DragonPage';

describe('Reducer tests', () => {
  test('Reducer return empty state if action is undefined', () => {
    const initial = [{dragon_id: '1', dragon_name: 'dragon1', dragon_type: 'type', dragon_description: 'description', dragon_image: 'image', reserved: false}];
    const action = () => ({type: 'unknown'});
    expect(actions.default(initial, action)).toEqual(initial);
  });
  test('Reserve Action converts to true reserved key', () => {
    const initial = [{dragon_id: '1', dragon_name: 'dragon1', dragon_type: 'type', dragon_description: 'description', dragon_image: 'image', reserved: false}];
    const final = [{dragon_id: '1', dragon_name: 'dragon1', dragon_type: 'type', dragon_description: 'description', dragon_image: 'image', reserved: true}];
    expect(actions.default(initial, actions.reserve(initial[0].dragon_id))).toEqual(final);
  });
  test('Cancel action converts to false reserved key', () => {
    const initial = [{dragon_id: '1', dragon_name: 'dragon1', dragon_type: 'type', dragon_description: 'description', dragon_image: 'image', reserved: true}];
    const final = [{dragon_id: '1', dragon_name: 'dragon1', dragon_type: 'type', dragon_description: 'description', dragon_image: 'image', reserved: false}];
    expect(actions.default(initial, actions.cancel(initial[0].dragon_id))).toEqual(final);
  })
})

describe('Body of the Dragons Page', () => {
  it('Check Dragon 1 appears in DOM', async () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <DragonPage />
        </Provider>
      </MemoryRouter>
    );
    await waitFor(() => {
      const text = screen.findByText(
        /Dragon is a reusable spacecraft developed by SpaceX, an American private space transportation company based in Hawthorne, California. Dragon is launched into space by the SpaceX Falcon 9 two-stage-to-orbit launch vehicle. The Dragon spacecraft was originally designed for human travel, but so far has only been used to deliver cargo to the International Space Station (ISS)./
      );
      expect(text).not.toBeNull();
    });
  });
  it('Matches snapshot', () => {
    const Tree = renderer.create(
      <MemoryRouter>
        <Provider store={store}>
          <DragonPage />
        </Provider>
      </MemoryRouter>
    ).toJSON();
    expect(Tree).toMatchSnapshot();
  });
});
