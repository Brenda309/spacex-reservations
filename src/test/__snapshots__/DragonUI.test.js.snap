// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body of the Dragons Page Matches snapshot 1`] = `
<div>
  <main
    className="container"
  >
    <section />
  </main>
</div>
`;
